### DISCUSSION TITLE
What would you like to see in a tutorial?

### DISCUSSION THREAD

**Discussion by XZoro-404** [kyleskom/NBA-Machine-Learning-Sports-Betting#455] [General]
Hello, I am making a video tutorial on how to setup this project on your local machine. Currently I have plans to make one for Windows 11 and Windows 10 machines. I would also be able to make one for some major linux distros if people would think it necessary.
If anyone has suggestions please drop a comment so I know what some of the pain points are for users.

**Comment by XZoro-404** [comment]
I am also considering creating a Jupyter notebook for testing and understanding purposes. However, there's no guarantee it would remain up to date with the code, as maintaining it could introduce an extra step for contributors. That said, I believe it could still be beneficial, especially since it has the potential to be run on a Google Colab Notebook.

**Comment by Tanstones630** [comment]
Hello,I haven't learned Python yet.  I have been reviewing the code of this project in the last two days.  I would like to offer a few suggestions：
1.Can you give the flowchart of this project. 
2.Explain the key code.

TeamData  dont' have 24-25 season's data, I need to update it  by step ?
cd src/Process-Data
1.python -m Get_Data
2.python -m Get_Odds_Data
3.python -m Create_Games
cd ../Train-Models
4.python -m XGBoost_Model_ML
5.python -m XGBoost_Model_UO

**Reply by XZoro-404** [reply]
This would be perfect! I might make this into a separate video from the install video since the main topic seems to be different enough.

**Comment by Youngstud4321** [comment]
When will the tutorial be posted?

**Reply by XZoro-404** [reply]
I was waiting for possible other suggestions, but I will have it out this weekend. I will post the link here as well as; I will try to get it onto the Readme.

**Reply by Youngstud4321** [reply]
any update on when the video will be posted?

**Reply by XZoro-404** [reply]
Yes my apologies I was on vacation for a while so I never had the chance to edit and upload the video but when I will have it posted this week. It will include both Windows 10 and Windows 11

**Reply by Youngstud4321** [reply]
Could you possibly add how to test the models with past games?

**Comment by glitched-coder** [comment]
Hello, Thank you for considering putting this together, excited to see the end product!

To be honest - I'm not a huge sports better. I kind of stumbled into this repository.  I just enjoy the world of AI and learning.

After digging through all the code - where I'm struggling to make connections is around the data acquisition.  Can't seem to wrap my head around the queries to the NBA website, how it's pulling that data, how it's segregating and organizing it, and how the Config.toml plays into all this.

I attempted to update this to pull start of 24-25 season to date data, but can't seem to get it working.  It gets stuck in the while loop within Get_Data.py.

The ML logic and how that's converging makes sense to me, I'm struggling with the data side and think that'd be helpful to walk through :)

**Reply by etonyetang** [reply]
You'll need to update the config.toml file to including the 24-25.
For example, 
   [create-games.2024-25]
        start_date = "2024-6-24"
        end_date = "2025-1-1"
        start_year = "2024"
        end_year = "2025"
Please help correct the above format. 
Also, needed to update file Create_Games.py and Add_Days_Rest.py under Add_Days_Rest.py
-Happy New Year's 2025.

**Comment by bklynate** [comment]
Any word on the tutorial?

**Reply by XZoro-404** [reply]
Yes, sorry I was dealing with a lot of vm errors (since I wanted fresh installs of windows) but since then I have now decided on trying to make it work through docker. This way it should run the same on windows/linux/mac

**Reply by nealmick** [reply]
Yea virtualization tends to be an issue with machine learning libraries as most libraries are written to be run directly on the host hardware.   You may look up bare metal server rental or dedicated hosts where the os is running directly on the hardware without virtualization i know aws ec2 has an option for this.

**Reply by XZoro-404** [reply]
I was able to run it on the VMs but since then I’ve been looking into Dockerizing this project, which would massively improve the user installation process and make it “OS agnostic”. 
The main concerns with Dockerizing this project would pose the following questions:
How would it affect performance?
What all needs to change to in the code to better conform to Docker?
Should the dockerized version be its own branch?

and I’m sure there are more such as, how long will this project be a 1 contributor project? Does the owner of this project want it to become a bigger thing? 
these questions don’t necessarily relate to Dockerizing the project but I do believe are important questions to the overall health of the project.

**Reply by nealmick** [reply]
Adding a dockerfile is a good idea, I dont think changes should be needed, hopefully should run as long as the packages are installed, I wouldn't worry too much about performance.   I normally only use docker to do deployments, not my favorite for running local containers.   Its not really 1 contributor people over time have definitely added stuff, i think the whole sb scrape data provider is maintained by [nkgilley](https://github.com/nkgilley).   

There was a youtuber named Siraj Raval who made a few videos regarding sports betting and used this repo in one of his videos which lead to substantial engagement [here](https://youtu.be/IDthta5sUGQ?si=6XZmfy7KlsX6m_dG&t=1098).  This youtuber originally i believe had good intentions and made educational videos on all sorts of subjects related to technology, software, crypto, and in particular AI.   His top videos have millions of views, and i was supportive at [first](https://github.com/kyleskom/NBA-Machine-Learning-Sports-Betting/discussions/106#discussioncomment-5053817).   However Siraj is now a scammer, here is [coffee zilla expose](https://www.youtube.com/watch?v=ovOaRSLTh1Q) there are multiple.  Siraj's video displays him using this project and other projects including his own to make money sports betting, as well as crpyto trading and so forth.  However its important to note that he displays a false narrative that you can make money by gambling with gpt or ai, or in other methods such as crypto trading ai bots.   I think this project started off as kyles personal adventure into machine learning.   Im in support of adding features and improving the project, I think adding more information making things clearer and easier for people to understand with tutorials and other content to help learn about the technical side would be good.   I just wanted to share some history of what i know, and like everything its not black and white, there is definitely some truth in siraj's video its not all one way or another.

**Reply by XZoro-404** [reply]
The one main thing that I haven’t been able to exactly figure out is running the main.py and then the app.py within the dockerfile. I found a temporary hack for it with 

> CMD ["sh", "-c", "python /app/main.py -xgb -odds=fanduel && cd Flask && flask run --debug"]

but this is clearly not the correct way to do it. 
I know docker won’t be for everyone but for anyone using windows without wanting to sink time in trying to fix errors just trying to run this project docker would be a tremendous savior. 
As for the 1 contributor, I might have not used the correct terminology but what I was meaning was 1 person to review and approve pushes.

**Reply by nealmick** [reply]
The docker command seems fine, i mean its kinda doing the same thing twice because the flask app runs `main.py -xgb -odds=fanduel ` here https://github.com/kyleskom/NBA-Machine-Learning-Sports-Betting/blob/9313a6c7da263c17286e0175d8812c02f1f91bdf/Flask/app.py#L24

I think the first part is running the shell but I doubt that's needed.  I mean to be honest its not clear what the user wants to do at first.... maybe they want to train a model or run the cli and not the web app... But yea the docker cmd looks good to me.   Maybe try something like this 
`WORKDIR /app/Flask `
`CMD ["flask", "run", "--host=0.0.0.0", "--port=8080"]`

**Reply by XZoro-404** [reply]
Just getting back into this after a busy schedule and during the time I've been away I have been learning more about flask and docker.
My new proposed solution to 
>  I mean to be honest its not clear what the user wants to do at first....
would be to always on first run grab new data (and of course the user would be able to change this by editing the DockerFile) then in the flask app we would be able to add options to to train models or alternatively grab new data. In order to not lock up the flask app during this time we could utilize celery workers to run the user requested process.

**Reply by nealmick** [reply]
Looks like you have big plans, it would definitely be cool to have model training interface with different settings and maybe even live training logs on the web page.  Im not sure if celery is necessary for background processes, It would work but adds a good bit of complexity with a MQ service or Redis.   I know when using Gunicorn in production it can be set with multiple workers and could be an easier solution for parallel processing.  I think celery really shines for distributed computing when you have multiple machines completing tasks in parallel, but i think since this is probably gona be all on the same machine then there are a few libraries designed for this, i found [this](https://stackoverflow.com/questions/22615475/flask-application-with-background-threads/22617121#22617121) stack overflow has a few potential solutions.

**Comment by t-dmac** [comment]
I tried this but then got an error when it tried to read one of the SQL database files. i checked the SQL lite and it was empty. how did you get around that ?

〽️ac

**Comment by bklynate** [comment]
@nealmick can you walk through what getting up and running with the repo from a fresh clone entails?

Do we need to gather new data initially? Do you run the models for training? Where can you look you start jumping in and possibly making changes to enhance things?

**Reply by nealmick** [reply]
The main setup needed to get the repo running is the python virtual environment with the required packages.   I would say the first step is getting the environment working then secondly see if you can run the main XGB model with `python3 main.py -xgb -odds=fanduel` or the flask app with `cd Flask flask --debug run`.   
To run the app you do not need to create a new dataset or fetch new training data.  `python -m Get_Data ` fetches the whole training dataset over again starting from 2007, it takes days to run.   `python -m Get_Odds_Data` does the same as Get Data but for odds.  `python -m Create_Games` combines and formats the odds and team data to create the finished dataset.   Creating a new dataset does make the data training cutoff closer to the current date.   Like with LLM's they have no knowledge after a certain date.  However the models are built to generalize and the game of basketball changes very slowly.   At the end of the season the same game is still being played, and the same model will have a similar accuracy, possibly after several seasons a model may start to become less accurate.  Also the repo has a training dataset that is already formatted and ready to train on.  

To train a model on the existing dataset the commands are `python -m XGBoost_Model_ML` and `python -m XGBoost_Model_UO`.  These commands use the existing dataset to train a new model.  The models are stored in the /Models/ folder with names like `XGBoost_53.7%_UO-9.json`.  To use a model you have trained you must updated the model runners with the path to the new model weights.  

The prediction system operates in two distinct phases: model training and daily predictions. The pre-trained models remain static, but the prediction process is dynamic and uses fresh data every day.

When you launch the Flask app or XGB model, it follows these steps:

1. Loads the pre-trained models from storage
2. Fetches today's game data from various data providers
3. Formats this new data to match the model's expected input
4. Generates predictions using the current data

A common misconception is that you need to constantly retrain the model with the latest games. This isn't necessary. While the model might be trained on historical data (for example, up through last season), it makes predictions using current team statistics and odds.
When making predictions, the model looks at each team's current stats, such as win/loss records and other relevant metrics. It doesn't need direct knowledge of recent games – their impact is already reflected in the updated team statistics it uses as input. The model can effectively generate predictions for today's games using these current statistics, even though it wasn't trained on yesterday's or last week's games.

**Comment by jtmeyers777** [comment]
This script looks amazing. I am quite tech savvy and know java programming but not much python. Would love a full walk through on setting this up. The degenerate gambler in me would love to have some comparison notes.

Have you ever thought of publishing results on a google worksheet for the simple minded folks that cant set this up?

**Comment by XZoro-404** [comment]
I think using python-rq would probably be the best as it seems to be more
lightweight on top of being a bit more developer friendly. All of the other
suggestions on that thread don't seem more outdated. In regards to having
to run Redis for it to work we could utilize docker compose to spin up all
required containers i.e. Web App and Redis (maybe python-rq I'm not sure
how to run it just yet)

**Reply by nealmick** [reply]
That sounds like a good option, Ive run celery a few times but it does kinda make for a more complex setup especially trying to debug.  I have used similar packages like python-rq for django which is basically like flask but more features.   I would co author a pr if you want to work on something together. we could even migrate to django, then we would have user profiles and a better structure for the web app.  Also Django has a sqlite ORM so we could connect to the database and create Django views to get data and display it from the database.  also on my last PR I included bootstrap on the html base template for the modal popup windows, which has alot of great stuff like sliders and responsive elements, can use them for more features like model training settings.
